---
title: "R Notebook"
output: html_notebook
---

Runs DBSCAN.

```{r}
library(dbscan)
library(tidyr)
library(readxl)

## Collate all baseline variant data into matrix
variant_data <- read_excel("C:/Users/Daniel/Documents/University/GNA5900/combined_variants.xlsx")
CNV_data <- read_excel("C:/Users/Daniel/Documents/University/GNA5900/combined_cnvs.xlsx")

combined_variants <- bind_rows(variant_data, CNV_data) %>%
  filter(!grepl("1\\.1|1\\.2|1\\.3|2\\.0|-1-1", sample)) %>%
  mutate(sample = sub("^([A-Za-z]+-[0-9]{3}).*", "\\1", sample)) %>%
  filter(plp == "Y")
  
variant_count <- combined_variants %>%
  group_by(gene, sample) %>%
  summarise(VariantCount = n(),.groups = "drop")

variant_matrix_hcluster <- variant_count %>%
  pivot_wider(
    names_from = sample,
    values_from = VariantCount,
    values_fill = 0
  )

variant_matrix_hcluster <- as.matrix(variant_matrix_hcluster[ , -1])
rownames(variant_matrix_hcluster) <- variant_count %>% distinct(gene) %>% pull(gene)

variant_matrix_hcluster <- t(variant_matrix_hcluster) %>%
  na.omit() %>%
  scale()

variant_matrix_hcluster <- as.matrix(variant_matrix_hcluster[ , -1]) # Exclude column of gene names

## Find appropriate DBSCAN parameters
kNNdistplot(variant_matrix_hcluster, minPts = 5)

dbscan_result <- dbscan(variant_matrix_hcluster, eps = 0.5)

## Display alignments
# print(dbscan_result$cluster)
# variant_matrix_hcluster$DBSCAN_Cluster <- as.factor(dbscan_result$cluster)

# Table
sample_ids <- rownames(variant_matrix_hcluster)
table(dbscan_result$cluster, sample_ids)

# Plotting Cluster
plot(dbscan_result, variant_matrix_hcluster, main = "DBScan")
plot(dbscan_result, variant_matrix_hcluster)
```

